### DISCUSSION TITLE
Window navigation - Jump to a numbered window

### DISCUSSION THREAD

**Discussion by nicholascross** [ghostty-org/ghostty#3950] [Feature Requests, Ideas]
This is tangentially related to #3193. I wasn't able to find a nice way to navigate between windows using the keyboard.

Rather than jump to next/previous windows it would be nice to jump to a numbered window using a key bind such as `cmd + <number>`.  I use this feature regularly in my current terminal emulator and am missing it in Ghostty.

**Comment by mitchellh** [comment]
How does the window get its number?

**Reply by nicholascross** [reply]
When a window is created the number is assigned.

So the first window opened is 1 and the subsequent is 2.

If we opened a third and closed the second.  The third window would still be referenced via the number 3 even though there are now only two open windows.

Opening another window at this point would assign it number 2 this is because we obviously can't increment numbers forever since there is only a certain number of single digit numbers on the keyboard 😄

**Reply by msr** [reply]
Big +1. On macOS, most terminals offer a keyboard shortcut to switch to a specific window:

- iTerm2 uses opt-command-# to switch windows by default.
- Terminal.app uses command-# to switch tabs by default, but can be configured to switch windows instead (in General settings).
- Kitty has https://sw.kovidgoyal.net/kitty/actions/#action-nth_os_window
- Wezterm has https://wezfurlong.org/wezterm/config/lua/keyassignment/ActivateWindow.html

iTerm2 & Terminal.app assign window numbers as @nicholascross describes: lowest unused when the window is created, numbers can be reused, no renumbering on closure. Kitty & Wezterm do it differently: windows always get a new number, numbers aren't reused, and `nth_os_window n` / `ActivateWindow(n)` switch to the open window with the n-th lowest number.

I personally prefer the iTerm2 / Terminal.app approach because it avoids changing a window's keybinding. With the Kitty / Wezterm approach, closing a window means the bindings change for all subsequent windows, which seems odd.

This would combine well with #4096 – make the window number a variable that can be shown in the title.

**Comment by canTuncPekkan** [comment]
I know that this is not what you want but it can be a temporary fix
you can try cmd + ~ this is bult-in macOS shortcut that acts like cmd + tab but for windows of an application.

if you have multiple ghostly windows open  cmd + ~ will go to the next one and shift + cmd + ~ will go to the previous one.

edit: the command was not cmd + ~ but cmd + `

**Reply by jrouzierinverse** [reply]
This does not work

**Reply by nicholascross** [reply]
The suggested approach does work for me but as the author wrote it is not exactly what I was after.  I often use this in a multi monitor set up with two main terminal windows open.  So this is not unhelpful for that case but as soon as you open anymore you can no longer **directly** navigate with a single set of key presses.

**Reply by jrouzierinverse** [reply]
Being able to target windows with `cmd+<number>` is a feature I would also like to have.
But `cmd + ~`  does not work for me for all with Ghostty which version are you running are you linux or Mac?

**Reply by canTuncPekkan** [reply]
I use macOS, I am very sorry I just looked at the keybind from the `system settings` app. it says that the shortcut was the uncapped version of my ~ key. I can't write it as markdown uses it as a format indicator of code blocks and I don't know how to type it normally in markdown.

to find the keybind in your settings follow; 
```
System settings > keyboard > keyboard shortcuts... > Keyboard > move focus to next window
```

**Reply by canTuncPekkan** [reply]
also this is not the solution the author wants, just a temporary alternative.

**Reply by jrouzierinverse** [reply]
Sorry for the confusion `cmd + ~ ` does work but not in full screen mode which is how I usually have my windows.

**Comment by DivineDominion** [comment]
As for numbering -- a battle-tested system is known from jump-to-thing-on-screen from

- `vim-easymotion` https://github.com/easymotion/vim-easymotion
- `avy` on Emacs https://github.com/abo-abo/avy

which are adapted at least to Emacs's `ace-window`: https://github.com/abo-abo/ace-window

1. Look at the window you want to go to
2. Press the 'jump to' shortcut
3. All windows show a letter/number/... overlay, very large and clearly visible
4. You already look where you want to go, read the letter/number/... and press it

Opposed to window numbering, this uses clear on screen "go here with this key" instructions.

<img width="1258" height="966" alt="image" src="https://github.com/user-attachments/assets/00656765-c1e3-404d-a7b6-0e13f6411da5" />

**Reply by rhodes-b** [reply]
I think you are in the wrong discussion windows are if you run 2 separate instances of ghostty each one of those is a window, each window can have its own number of tabs and splits (what you are showing in that image is splits)
you can also get separate windows if you pull a tab out into its own window its not required to be a separate process

The behavior of supporting showing numbers for selecting a split is this one
https://github.com/ghostty-org/ghostty/discussions/3212

**Reply by DivineDominion** [reply]
Ah, there are great examples already, too. Nice!

I'd also like to switch windows, not just panes, esp. because once two windows exist, we can't merge them. But the pane switching maybe should come first then.

**Comment by smenor** [comment]
Just gonna +1 on the cmd+# ( easymotion is also fine but *something* would be exctremely appreciated )

**Comment by jrouzierinverse** [comment]
This should be closed since it is working with tabs within a window

**Comment by softwareentrepreneer** [comment]
+1
